frontend/src/service/ChangeLog.service.ts

```diff
@@ -1,73 +1,66 @@
-import { Update } from '../entity/Update';
-
-type Time = number;
+import { ChangeLog } from '../entity/ChangeLog';
+import { ChangeLogGraphQLApi } from './ChangeLogGraphQL.api';
+import { ErrorService, Err } from './Error.service';
 
 export class ChangeLogService {
+  constructor(
+    private changeLogGraphQLApi: ChangeLogGraphQLApi,
+    private errorService: ErrorService
+  ) {}
+
   hasUpdates(): Promise<boolean> {
     return new Promise(async (resolve, reject) => {
-      const changeLog = await this.getChangeLog();
-      const lastSeenAt = await this.getLastSeenChangeLog();
-      if (!changeLog) {
-        resolve(false);
-        return;
-      }
-
-      if (!changeLog[0]) {
-        resolve(false);
-        return;
-      }
+      try {
+        const changeLog = await this.changeLogGraphQLApi.getChangeLog();
+        if (!changeLog.changes) {
+          resolve(false);
+          return;
+        }
 
-      if (lastSeenAt >= changeLog[0].releasedAt) {
-        resolve(false);
-        return;
-      }
+        if (!changeLog.changes[0]) {
+          resolve(false);
+          return;
+        }
 
-      resolve(true);
-    });
-  }
+        if (!changeLog.lastViewedAt) {
+          resolve(true);
+          return;
+        }
 
-  getLastSeenChangeLog(): Promise<Time> {
-    return new Promise(async (resolve, reject) => {
-      resolve(await this.invokeLastSeenChangeLogApi());
-    });
-  }
+        for (let i = 0; i < changeLog.changes.length; i++) {
+          if (
+            new Date(changeLog.lastViewedAt).getTime() <
+            new Date(changeLog.changes[i].releasedAt).getTime()
+          ) {
+            resolve(true);
+            return;
+          }
+        }
```

Comment:
You can do a simple [map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) and find the most recent releaseAt. Then compare it with `lastViewedAt`. Separate the logics not only improves readability but also allow us to optimize the algorithm in the future.

Reply:
We will have to sort the changes based on timestamp when displaying them eventually.

Reply:
We can create a helper method in this class to fetch the change log and sort them.